topology/graph/processor.go

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2018 Orange
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package graph
+
+import (
+	"github.com/skydive-project/skydive/common"
+)
+
+// NodeAction is a callback to perform on a node. The action is kept
+// active as long as it returns true.
+type NodeAction interface {
+	ProcessNode(g *Graph, n *Node) bool
+}
+
+// deferred represents a node action with additional info if needed
+// for cancellation.
+type deferred struct {
+	action NodeAction
+}
+
+// Processor encapsulates an indexer that will process NodeActions
+// on the nodes that filter
+type Processor struct {
+	common.RWMutex
+	*MetadataIndexer
+	actions map[string][]deferred
+}
+
+// NewProcessor creates a Processor on the graph g, a stream of
+// events controlled by listenerHandler, that match a first set
+// of metadata m. Actions will be associated to a given set
+// of values for indexes.
+func NewProcessor(g *Graph, listenerHandler ListenerHandler, m ElementMatcher, indexes ...string) (processor *Processor) {
+	processor = &Processor{
+		MetadataIndexer: NewMetadataIndexer(g, listenerHandler, m, indexes...),
+		actions:         make(map[string][]deferred),
+	}
+	processor.MetadataIndexer.AddEventListener(processor)
+	return
+}
+
+// DoAction willl perform the action for nodes matching
+func (processor *Processor) DoAction(action NodeAction, values ...interface{}) {
```

Comment:
Maybe calling it "RegisterAction" would make it clearer that the action will be repeated. What do you think ?